File: VIF_git/GSDS-Qual/GiChul/23-2/1.c
#include <stdio.h>
#include <stdlib.h>

struct Node{
    char id;
    struct Node *adj[5];
    int adjlen;
};
    
int bfs(struct Node *start){
    struct Node *queue[6];
    queue[0] = start;
    int qlen = 1;
    char visited[6];
    int vlen = 0;

    while(qlen!=0){
        struct Node *curr_node = queue[qlen-1];
        qlen--;

        for(int i = 0; i < curr_node->adjlen; i++){
            struct Node *next_node = curr_node->adj[i];

            int isin = 0;
            for(int j = 0; j < vlen; j++){
                if(visited[j]==next_node->id) isin=1;
            }

            if(isin==0){
                visited[vlen]=next_node->id;
                vlen++;
                queue[qlen]=next_node;
                qlen++;
            }
        }
    }
    
    return vlen;
}

void find_root(struct Node *G[], int Glen){
    char res[6];
    int reslen = 0;
    for(int i = 0; i < Glen; i++){
        struct Node *start_node = G[i];
        if(bfs(start_node) == Glen){
            res[reslen]=start_node->id;
            reslen++;
        }
    }

    printf("[");
    for(int i = 0; i < reslen; i++){
        printf("%c",res[i]);
        if(i!=reslen-1) printf(", ");
    }
    printf("]\n");
}

int main(){
    struct Node A, B, C, D, E, F;
    A.id = 'A';
    B.id = 'B';
    C.id = 'C';
    D.id = 'D';
    E.id = 'E';
    F.id = 'F';
    A.adj[0] = &C;
    A.adj[1] = &D;
    B.adj[0] = &A;
    B.adj[1] = &F;
    C.adj[0] = &D;
    C.adj[1] = &B;
    D.adj[0] = &E;
    E.adj[0] = &F;
    A.adjlen = 2;
    B.adjlen = 2;
    C.adjlen = 2;
    D.adjlen = 1;
    E.adjlen = 1;
    F.adjlen = 0;

    struct Node *G[6] = {&A,&B,&C,&D,&E,&F};

    find_root(G,6);

    return 0;
}

File: VIF_git/GSDS-Qual/GiChul/23-2/1.py
class GNode:   
    def __init__(self, id, color='W', p=None):        
        self.id = id        
        self.color = color        
        self.parent = p            
    
    def __str__(self):        
        return self.id     
       
    def __hash__(self):        
        return hash(self.id)
    
A, B, C = GNode('A'), GNode('B'), GNode('C')
D, E, F = GNode('D'), GNode('E'), GNode('F')
    
G = dict()
    
G[A], G[B], G[C] = [C,D], [A,F], [D,B]
G[D], G[E], G[F] = [E], [F], []


def bfs(start_node, graph):
    queue = [start_node]
    visited = [start_node]

    while queue:
        curr_node = queue.pop(0)

        for next_node in graph[curr_node]:
            if next_node not in visited:
                visited.append(next_node)
                queue.append(next_node)
    return visited

def find_root(G):
    res = []
    for start_node in G.keys():
        if len(bfs(start_node, G)) == len(G.keys()): res.append(start_node.id)

    return res

print(find_root(G))


